distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java

```diff
@@ -107,11 +107,18 @@ public void execute(Terminal terminal, OptionSet options, Environment env, Proce
 
         // we are running in the foreground, so wait for the server to exit
         int exitCode = server.waitFor();
+        onExit(exitCode);
         if (exitCode != ExitCodes.OK) {
```

Comment:
nit: can we move throwing this UserException to the new method? Then serverless can just call super.